File: packages/expo-camera/android/src/main/java/expo/modules/camera/CameraViewManager.java
package expo.modules.camera;

import android.content.Context;

import com.google.android.cameraview.AspectRatio;
import com.google.android.cameraview.Size;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.unimodules.core.ViewManager;
import org.unimodules.core.interfaces.ExpoProp;
import org.unimodules.core.ModuleRegistry;
import org.unimodules.core.interfaces.ModuleRegistryConsumer;
import org.unimodules.core.interfaces.services.UIManager;
import org.unimodules.interfaces.barcodescanner.BarCodeScannerSettings;

public class CameraViewManager extends ViewManager<ExpoCameraView> implements ModuleRegistryConsumer {
  public enum Events {
    EVENT_CAMERA_READY("onCameraReady"),
    EVENT_ON_MOUNT_ERROR("onMountError"),
    EVENT_ON_BAR_CODE_SCANNED("onBarCodeScanned"),
    EVENT_ON_FACES_DETECTED("onFacesDetected"),
    EVENT_ON_FACE_DETECTION_ERROR("onFaceDetectionError"),
    EVENT_ON_PICTURE_SAVED("onPictureSaved");

    private final String mName;

    Events(final String name) {
      mName = name;
    }

    @Override
    public String toString() {
      return mName;
    }
  }

  private static final String REACT_CLASS = "ExponentCamera";
  private ModuleRegistry mModuleRegistry;

  @Override
  public void setModuleRegistry(ModuleRegistry moduleRegistry) {
    mModuleRegistry = moduleRegistry;
  }

  @Override
  public void onDropViewInstance(ExpoCameraView view) {
    mModuleRegistry.getModule(UIManager.class).unregisterLifecycleEventListener(view);
    view.stop();
  }

  @Override
  public String getName() {
    return REACT_CLASS;
  }

  @Override
  public ViewManagerType getViewManagerType() {
    return ViewManagerType.GROUP;
  }

  @Override
  public ExpoCameraView createViewInstance(Context context) {
    return new ExpoCameraView(context, mModuleRegistry);
  }

  @Override
  public List<String> getExportedEventNames() {
    List<String> eventNames = new ArrayList<>(Events.values().length);
    for(Events event : Events.values()) {
      eventNames.add(event.toString());
    }
    return eventNames;
  }

  @ExpoProp(name = "type")
  public void setType(ExpoCameraView view, int type) {
    view.setFacing(type);
  }

  @ExpoProp(name = "ratio")
  public void setRatio(ExpoCameraView view, String ratio) {
    view.setAspectRatio(AspectRatio.parse(ratio));
  }

  @ExpoProp(name = "flashMode")
  public void setFlashMode(ExpoCameraView view, int torchMode) {
    view.setFlash(torchMode);
  }

  @ExpoProp(name = "autoFocus")
  public void setAutoFocus(ExpoCameraView view, boolean autoFocus) {
    view.setAutoFocus(autoFocus);
  }

  @ExpoProp(name = "focusDepth")
  public void setFocusDepth(ExpoCameraView view, float depth) {
    view.setFocusDepth(depth);
  }

  @ExpoProp(name = "zoom")
  public void setZoom(ExpoCameraView view, float zoom) {
    view.setZoom(zoom);
  }

  @ExpoProp(name = "whiteBalance")
  public void setWhiteBalance(ExpoCameraView view, int whiteBalance) {
    view.setWhiteBalance(whiteBalance);
  }

  @ExpoProp(name = "pictureSize")
  public void setPictureSize(ExpoCameraView view, String size) {
    view.setPictureSize(Size.parse(size));
  }

  @ExpoProp(name = "barCodeScannerSettings")
  public void setBarCodeScannerSettings(ExpoCameraView view, Map<String, Object> settings) {
    view.setBarCodeScannerSettings(new BarCodeScannerSettings(settings));
  }

  @ExpoProp(name = "useCamera2Api")
  public void setUseCamera2Api(ExpoCameraView view, boolean useCamera2Api) {
    view.setUsingCamera2Api(useCamera2Api);
  }

  @ExpoProp(name = "barCodeScannerEnabled")
  public void setBarCodeScanning(ExpoCameraView view, boolean barCodeScannerEnabled) {
    view.setShouldScanBarCodes(barCodeScannerEnabled);
  }

  @ExpoProp(name = "faceDetectorEnabled")
  public void setFaceDetectorEnabled(ExpoCameraView view, boolean faceDetectorEnabled) {
    view.setShouldDetectFaces(faceDetectorEnabled);
  }

  @ExpoProp(name = "faceDetectorSettings")
  public void setFaceDetectorSettings(ExpoCameraView view, Map<String, Object> settings) {
    view.setFaceDetectorSettings(settings);
  }
}

File: packages/expo-camera/android/src/main/java/expo/modules/camera/CameraViewManager.kt
package expo.modules.camera

import com.google.android.cameraview.AspectRatio
import com.google.android.cameraview.Size
import expo.modules.core.interfaces.services.UIManager
import expo.modules.interfaces.barcodescanner.BarCodeScannerSettings
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class CameraViewModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExponentCamera")

    View(ExpoCameraView::class) {
      Events(
        "onCameraReady",
        "onMountError",
        "onBarCodeScanned",
        "onFacesDetected",
        "onFaceDetectionError",
        "onPictureSaved"
      )

      OnViewDestroys<ExpoCameraView> { view ->
        val uiManager = appContext.legacyModule<UIManager>()
        uiManager?.unregisterLifecycleEventListener(view)
        view.cameraView.stop()
      }

      Prop("type") { view: ExpoCameraView, type: Int ->
        view.cameraView.facing = type
      }

      Prop("ratio") { view: ExpoCameraView, ratio: String? ->
        if (ratio == null) {
          return@Prop
        }
        view.cameraView.setAspectRatio(AspectRatio.parse(ratio))
      }

      Prop("flashMode") { view: ExpoCameraView, torchMode: Int ->
        view.cameraView.flash = torchMode
      }

      Prop("autoFocus") { view: ExpoCameraView, autoFocus: Boolean ->
        view.cameraView.autoFocus = autoFocus
      }

      Prop("focusDepth") { view: ExpoCameraView, depth: Float ->
        view.cameraView.focusDepth = depth
      }

      Prop("zoom") { view: ExpoCameraView, zoom: Float ->
        view.cameraView.zoom = zoom
      }

      Prop("whiteBalance") { view: ExpoCameraView, whiteBalance: Int ->
        view.cameraView.whiteBalance = whiteBalance
      }

      Prop("pictureSize") { view: ExpoCameraView, size: String? ->
        if (size == null) {
          return@Prop
        }
        view.cameraView.pictureSize = Size.parse(size)
      }

      Prop("barCodeScannerSettings") { view: ExpoCameraView, settings: Map<String, Any?>? ->
        view.setBarCodeScannerSettings(BarCodeScannerSettings(settings))
      }

      Prop("useCamera2Api") { view: ExpoCameraView, useCamera2Api: Boolean ->
        view.cameraView.setUsingCamera2Api(useCamera2Api)
      }

      Prop("barCodeScannerEnabled") { view: ExpoCameraView, barCodeScannerEnabled: Boolean ->
        view.setShouldScanBarCodes(barCodeScannerEnabled)
      }

      Prop("faceDetectorEnabled") { view: ExpoCameraView, faceDetectorEnabled: Boolean ->
        view.setShouldDetectFaces(faceDetectorEnabled)
      }

      Prop("faceDetectorSettings") { view: ExpoCameraView, settings: Map<String, Any>? ->
        view.setFaceDetectorSettings(settings)
      }
    }
  }
}
